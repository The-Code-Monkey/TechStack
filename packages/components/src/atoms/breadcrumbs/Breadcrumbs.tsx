import { TypographyProps } from '@aw-web-design/styled-system';
import { useEffect, useState } from 'react';

import { Box, Text, BoxProps } from '../../primal';
import { Link } from '../link';

export interface Props extends BoxProps {
  homeUrl?: string;
  homeText?: string;
  url?: string;
  breadcrumbs?: Array<string>;
  typography?: TypographyProps;
}

const Breadcrumbs = ({
  homeUrl,
  homeText,
  url,
  breadcrumbs,
  typography,
  ...rest
}: Props) => {
  const [crumbs, setCrumbs] = useState([]);

  const getHref = index => {
    let href = '/';

    for (let i = 0; i < index + 1; i++) {
      href += `${i !== 0 ? '/' : ''}${crumbs[i]}`;
    }

    return href;
  };

  useEffect(() => {
    if (!breadcrumbs && !url && typeof window !== 'undefined') {
      setCrumbs(window.location.pathname.replace(/^\/|\/$/g, '').split('/'));
    } else if (breadcrumbs) {
      setCrumbs(breadcrumbs);
    } else if (url) {
      setCrumbs(url.replace(/^\/|\/$/g, '').split('/'));
    }
  }, [breadcrumbs, url, window]);

  return (
    <Box {...rest}>
      <Text {...typography}>
        <Link href={homeUrl ?? '/'}>{homeText ?? 'common.home'}</Link>
        {' > '}
        {crumbs.length > 0 &&
          crumbs
            .map((c, i) => (
              <Link key={c} href={getHref(i)}>
                {c}
              </Link>
            ))
            .reduce(
              (prev, curr) => [prev, ' > ', curr] as unknown as JSX.Element
            )}
      </Text>
    </Box>
  );
};

export default Breadcrumbs;
