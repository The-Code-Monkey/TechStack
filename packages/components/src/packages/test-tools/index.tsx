import { mount, shallow, ShallowWrapper, ReactWrapper } from 'enzyme';
import React, { ReactNode } from 'react';
import { ThemeProvider } from 'styled-components';

import theme from '../theme';
import { ConfigContext } from '../utils';

const config = {
  outputDir: '/src/packages/',
  iconDir: 'feather',
};

const removeProperties = () => {
  const keys = ['theme', 'styledTheme'];
  return {
    test: (val: any) =>
      val &&
      typeof val === 'object' &&
      'props' in val &&
      Object.keys(val.props).some((prop) => keys.some((key) => key === prop)),
    print: (val: any, serialize: any) => {
      keys.forEach((key) => {
        delete val.props[key];
      });
      return serialize(val);
    },
  };
};

const formattedTheme = {
  ...theme,
  colors: {
    common: theme.colors?.common,
    modes: theme.colors?.modes,
    ...theme.colors?.modes.light,
  },
};

export const shallowWithTheme: (children: ReactNode) => ShallowWrapper = (
  children: ReactNode
) =>
  shallow(<ThemeProvider theme={formattedTheme}>{children}</ThemeProvider>)
    .dive()
    .shallow();

export const mountWithTheme: (children: ReactNode) => ReactWrapper = (
  children: ReactNode
) => {
  expect.addSnapshotSerializer(removeProperties());

  return mount(
    <ConfigContext.Provider value={config}>
      <ThemeProvider theme={formattedTheme}>
        <>{children}</>
      </ThemeProvider>
    </ConfigContext.Provider>
  );
};
