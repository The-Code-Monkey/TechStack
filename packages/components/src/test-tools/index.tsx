import isPropValid from '@emotion/is-prop-valid';
import { render, RenderResult } from '@testing-library/react';
import { ReactNode } from 'react';
import { StyleSheetManager, ThemeProvider } from 'styled-components';
import { expect } from 'bun:test';

import theme from '../theme';
import { ConfigContext } from '../utils';

const config = {
  outputDir: '/src/packages/',
  iconDir: 'feather',
};

const removeProperties = () => {
  const keys = ['theme', 'styledTheme'];
  return {
    test: (val: { props: unknown }) =>
      val &&
      typeof val === 'object' &&
      'props' in val &&
      Object.keys(val.props).some(prop => keys.some(key => key === prop)),
    serialize: (val, config, indentation, depth, refs, printer) => {
      keys.forEach(key => {
        delete val.props[key];
      });
      return printer(val, config, indentation, depth, refs);
    },
  };
};

const formattedTheme = {
  ...theme,
  colors: {
    common: theme.colors?.common,
    modes: theme.colors?.modes,
    ...theme.colors?.modes.light,
  },
};

export const mountWithTheme = (children: ReactNode): RenderResult => {
  // expect.addSnapshotSerializer(removeProperties());

  return render(
    <StyleSheetManager shouldForwardProp={isPropValid}>
      <ConfigContext.Provider value={config}>
        <ThemeProvider theme={formattedTheme}>
          <>{children}</>
        </ThemeProvider>
      </ConfigContext.Provider>
    </StyleSheetManager>
  );
};
