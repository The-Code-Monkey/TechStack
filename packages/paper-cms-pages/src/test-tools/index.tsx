import { ConfigContext } from '@techstack/components';
import { render } from '@testing-library/react';
import { ReactNode } from 'react';

import theme from '../theme';
import { ThemeProvider } from '../workarounds/styled-components';

const config = {
  outputDir: '/src/packages/',
  iconDir: 'feather',
};

const removeProperties = () => {
  const keys = ['theme', 'styledTheme'];
  return {
    test: (val: { props: unknown }) =>
      val &&
      typeof val === 'object' &&
      'props' in val &&
      Object.keys(val.props).some(prop => keys.some(key => key === prop)),
    serialize: (val, config, indentation, depth, refs, printer) => {
      keys.forEach(key => {
        delete val.props[key];
      });
      return printer(val, config, indentation, depth, refs);
    },
  };
};

const formattedTheme = {
  ...theme,
  colors: {
    common: theme.colors?.common,
    modes: theme.colors?.modes,
    ...theme.colors?.modes.light,
  },
};

export const mountWithTheme = (children: ReactNode) => {
  expect.addSnapshotSerializer(removeProperties());

  return render(
    <ConfigContext.Provider value={config}>
      <ThemeProvider theme={formattedTheme}>
        <>{children}</>
      </ThemeProvider>
    </ConfigContext.Provider>
  );
};
