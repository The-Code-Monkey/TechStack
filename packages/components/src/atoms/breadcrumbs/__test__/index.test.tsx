import { mountWithTheme } from '../../../test-tools';
import Breadcrumbs from '../Breadcrumbs';

describe('<Breadcrumbs />', () => {
  it('should render correctly with url param', () => {
    const { asFragment } = mountWithTheme(
      <Breadcrumbs url='/test/test/help/dsdasd/' />
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
