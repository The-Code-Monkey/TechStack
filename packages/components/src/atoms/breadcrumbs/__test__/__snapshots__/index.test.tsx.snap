// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Breadcrumbs /> should render correctly with url param 1`] = `
<DocumentFragment>
  <div
    class="sc-gsnTZi cXXFdo"
    data-autoid="box"
  >
    <span
      class="sc-eCYdqJ kQHdnO"
      data-autoid=""
    >
      <a
        href="/"
      >
        common.home
      </a>
       &gt; 
      <a
        href="/test"
      >
        test
      </a>
       &gt; 
      <a
        href="/test/test"
      >
        test
      </a>
       &gt; 
      <a
        href="/test/test/help"
      >
        help
      </a>
       &gt; 
      <a
        href="/test/test/help/dsdasd"
      >
        dsdasd
      </a>
    </span>
  </div>
</DocumentFragment>
`;
